contracts/asset-proxy/contracts/src/bridges/DexForwarderBridge.sol

```diff
@@ -0,0 +1,213 @@
+        bytes bridgeData;
+    }
+
+    /// @dev Intermediate state variables used by `bridgeTransferFrom()`, in
+    ///      struct form to get around stack limits.
+    struct TransferFromState {
+        address inputToken;
+        uint256 initialInputTokenBalance;
+        uint256 callInputTokenAmount;
+        uint256 callOutputTokenAmount;
+        uint256 totalInputTokenSold;
+        BridgeCall[] calls;
+    }
+
+    event DexForwarderBridgeCallFailed(
+        address indexed target,
+        address inputToken,
+        address outputToken,
+        uint256 inputTokenAmount,
+        uint256 outputTokenAmount
+    );
+
+    /// @dev Spends this contract's entire balance of input tokens by forwarding
+    /// them to other bridges. Reverts if the entire balance is not spent.
+    /// @param outputToken The token being bought.
+    /// @param to The recipient of the bought tokens.
+    /// @param bridgeData The abi-encoded input token address.
+    /// @return success The magic bytes if successful.
+    function bridgeTransferFrom(
+        address outputToken,
+        address /* from */,
+        address to,
+        uint256 /* amount */,
+        bytes calldata bridgeData
+    )
+        external
+        returns (bytes4 success)
+    {
+        TransferFromState memory state;
+        (
+            state.inputToken,
+            state.calls
+        ) = abi.decode(bridgeData, (address, BridgeCall[]));
+
+        state.initialInputTokenBalance = IERC20Token(state.inputToken).balanceOf(address(this));
+
+        for (uint256 i = 0; i < state.calls.length; ++i) {
+            // Stop if the we've sold all our input tokens.
+            if (state.totalInputTokenSold >= state.initialInputTokenBalance) {
+                break;
+            }
+
+            BridgeCall memory call = state.calls[i];
+            // Compute token amounts.
+            state.callInputTokenAmount = LibSafeMath.min256(
+                call.inputTokenAmount,
+                state.initialInputTokenBalance.safeSub(state.totalInputTokenSold)
+            );
+            state.callOutputTokenAmount = LibMath.getPartialAmountFloor(
+                state.callInputTokenAmount,
+                call.inputTokenAmount,
+                call.outputTokenAmount
+            );
+
+            // Execute the call in a new context so we can recoup transferred
+            // funds by reverting.
+            (bool didSucceed, ) = address(this)
+                .call(abi.encodeWithSelector(
+                    this.executeBridgeCall.selector,
+                    call.target,
+                    to,
+                    state.inputToken,
+                    outputToken,
+                    state.callInputTokenAmount,
+                    state.callOutputTokenAmount,
+                    call.bridgeData
+                ));
+
+            if (!didSucceed) {
+                // Log errors.
+                emit DexForwarderBridgeCallFailed(
```

Comment:
Is this just for internal data gathering/insights? 

AS makes heavy use of FillOrKill so the only case where it'd reasonably be emitted is when succeed with enough slippage/backup path?

Reply:
Yeah, that's what I'm thinking. These won't be going through `fillOrder()` or the ERC20BridgeProxy so it's a less obvious path to detecting them. Maybe it's not worth it, though, since the bridge itself can revert if not enough input tokens are sold. We'd have to rely on traces for those situations (and the market fn failing).

Now I'm leaning towards nixing this event.